lib/src/cli/deployapps/baas_client.dart

```diff
@@ -0,0 +1,302 @@
+      "resetPasswordSubject": "",
+      "resetPasswordUrl": "http://localhost/resetPassword",
+      "runConfirmationFunction": true,
+      "runResetFunction": true
+    }''');
+
+    await _createMongoDBService(app, '''{
+      "flexible_sync": {
+        "state": "enabled",
+        "database_name": "flexible_sync_data",
+        "queryable_fields_names": ["TODO"],
+        "permissions": {
+          "rules": {},
+          "defaultRoles": [
+            {
+              "name": "all",
+              "applyWhen": {},
+              "read": true,
+              "write": true
+            }
+          ]
+        }
+      }
+    }''');
+
+    await _put('groups/$_groupId/apps/$app/sync/config', '{ "development_mode_enabled": true }');
+
+    return app;
+  }
+
+  Future<void> enableProvider(BaasApp app, String type, [String config = '{}']) async {
+    print('Enabling provider $type for ${app.name}');
+
+    final url = 'groups/$_groupId/apps/$app/auth_providers';
+    if (type == 'api-key') {
+      final providers = await _get(url) as List<dynamic>;
+      final apiKeyProviderId = providers.singleWhere((dynamic doc) => doc['type'] == 'api-key')['_id'] as String;
+
+      await _put('$url/$apiKeyProviderId/enable', '{}');
+    } else {
+      await _post(url, '''{
+          "name": "$type",
+          "type": "$type",
+          "disabled": false,
+          "config": $config
+        }''');
+    }
+  }
+
+  Future<void> _authenticate(String provider, String credentials) async {
+    dynamic response = await _post('auth/providers/$provider/login', credentials);
+
+    _headers['Authorization'] = "Bearer ${response['access_token']}";
+  }
+
+  Future<String> _createFunction(BaasApp app, String name, String source) async {
+    print('Creating function $name for ${app.name}...');
+
+    final dynamic response = await _post('groups/$_groupId/apps/$app/functions', '''{
+        "name": "$name",
+        "source": ${jsonEncode(source)},
+        "private": false,
+        "can_evaluate": {}
+      }''');
+
+    return response['_id'] as String;
+  }
+
+  Future<String> _createMongoDBService(BaasApp app, String syncConfig) async {
+    final serviceName = _clusterName == null ? 'mongodb' : 'mongodb-atlas';
+    final mongoConfig = _clusterName == null ? '{ "uri": "mongodb://localhost:26000" }' : '{ "clusterName": "$_clusterName" }';
+    final mongoServiceId = await createService(app, 'BackingDB', serviceName, mongoConfig);
+
+    // The cluster linking must be separated from enabling sync because Atlas
+    // takes a few seconds to provision a user for BaaS, meaning enabling sync
+    // will fail if we attempt to do it with the same request. It's nondeterministic
+    // how long it'll take, so we must retry for a while.
+    var attempt = 0;
+    while (true) {
+      try {
+        await _patch('groups/$_groupId/apps/$app/services/$mongoServiceId/config', syncConfig);
+        break;
+      } catch (err) {
+        if (attempt++ < 1) {
```

Comment:
It will cause only 1 retry, if I'm not wrong. In the message is written "Will keep retrying". Is one retry enough?

Reply:
Well spotted - I changed it to make sure that the error gets propagated correctly and forgot to return it to a more sane value.